core/src/main/scala/kafka/log/Log.scala

```diff
@@ -180,6 +180,17 @@ object RollParams {
   }
 }
 
+sealed trait LogStartOffsetIncrementCause
+case object ClientRecordDeletion extends LogStartOffsetIncrementCause {
+  override def toString: String = "client delete records request"
+}
+case object LeaderOffsetIncremented extends LogStartOffsetIncrementCause {
+  override def toString: String = "leader offset increment"
+}
+case object SegmentDeletion extends LogStartOffsetIncrementCause {
+  override def toString: String = "segment deletion to enforce retention"
```

Comment:
Maybe we should just say segment deletion? We can also delete segments during recovery and such. Could that not trigger this path?